### ISSUE TITLE
Locating the weights of a trained model

### ISSUE THREAD

**Issue by juakesson** [MIC-DKFZ/nnUNet#372]
Hello!

I ran the entire nnU-Net process for my own dataset, and I'm now trying to find the weights of the trained model for each fold. However, I cannot seem to find where the weights are saved. I figure it should be in the model_best.model-file, but I don't really understand the format of this. Where can I find the weights and how do I load them into memory after training is finished?

Another question: Are the intermediate outputs that are used to inject gradients during training ignored during inference? In that case, where in the code can this be found?

My intention is to try to export the trained models via ONNX format, in order to be able to import them into MATLAB, using them as a baseline for further development. 

Thanks in advance!

**Comment by FabianIsensee** [comment]
Hi,
here is some of the information you need https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/common_questions.md#where-can-i-find-the-segmentation-metrics-of-my-experiments

This post is about the metrics, but the weights are also in the folder structure.

If you want to load the weights, initialize the trainer class (as in run_training.py or the function `load_model_and_checkpoint_files`) and the run the TRAINERCLASS.load_checkoint function. You can then access the network via TRAINERCLASS.network. 

The deep supervision is deactivated during inference. See `Generic_UNet.do_ds`.

If you export the model via ONNX for inference, make sure to set do_ds=False before you do the export. ONNX will then know nothing about the deep supervision.

You need to make sure the preprocessing in matlab is identical to the preprocessing in nnU-Net. Also implementing the inference scheme is not straightforward. You can't just plug whole cases into the network. You need to use the same patch size for inference that you also used during training. Whatever you do, please make sure the inference in Matlab gives the same results (you should run nnunet inference and matlab inference on some validation set and compare the results!)

Best,
Fabian

**Comment by juakesson** [comment]
Hello Fabian!

Thank you so much for your answer. It was very helpful. 

I have some follow-up questions regarding a few preprocessing details:

1. Reading the .pkl-file  nnUNetPlansv2.1_plans_2D.pkl, the inferred patch size for my (in-house MRI) dataset is given to be [320, 384], while the median_patient_size_in_voxels is [16, 298, 382]. The spacing of my images from the preprocessing seems to be unchanged. For the 2D-network, how is the sampling (and oversampling) using this patch size handled before training and how is the inference using the sliding window-approach performed for small images? Is padding performed on each patch for images smaller than the inferred patch size in order to get the correct patch size? I couldn't find information about this in your manuscript or code. If zero-padding is performed - what is your perception of how this affects the performance of the segmentation?

Also, besides using the correct patch size, is there any additional pre-processing done to test examples before inference? In that case, in which file can details on these be found?

Best regards, 
Julius

**Comment by FabianIsensee** [comment]
Hi Julius,
oversampling is implemented as follows:
1) select a random training case
2) select a random foreground class that exists in that training case
3) select one of the slices that contains it
4) select the patch in the slice such that it contains the desired class

If images are too small for a given patch size, they will be padded with zeroes, both during training and inference. This does not cause any issues with segmentation performance in my experience.

During testing, the same preprocessing will be applied that was also applied to the training images. This is saved in the plans file 
Best,
Fabian
